---
title: "InRelate - An R Package for Estimating Genetic Relatedness in Structured Populations"
author: "Khuyen Nguyen and Arun Sethuraman"
date: "2/17/2024"
output: html_document
---

# 1. Introduction

This vignette aims to provide users with a comprehensive tutorial on the fundamental useage of the InRelate package. The primary objective of this software is to precisely assess genetic relatedness while taking into consideration the impact of population structure. Many existing genetic relatedness estimators out in the field assume homogeneous populations and fail to account for the presence of population structure, therefore, this R package intends to fill this gap. InRelate utilizes multi-allelic population genomic data, considers nine different identify by descent (IBD) states, and implements a maximum likelihood estimator of pairwise genetic relatedness in structured populations. 

## 1.1 Installation

To install this package, you must download it from its designated GitHub repository (https://github.com/itskhuyen/InRelate). TIt is important to note that the CRAN update version is currently undergoing the submission process for integration into Rstudio. All questions should be addressed to Khuyen Nguyen or Dr. Sethuraman, the developers of the package.


To utilize the InRelate package, users must ensure that RStudio is installed on their laptop. Given that the CRAN version is not yet published, it is important to download the tar.gz file from the GitHub repository for execution on both Mac and Windows operating systems. For Mac and Windows users, follow these steps:

1. Open RStudio on your machine.
2. Navigate to the "Tools" menu.
3. Click on "Install Packages."
4. Under "Install FROM," choose "Package Archive File (.tar.gz)."
5. Select the file "InRelate_0.1.0.tar.gz" from your downloaded location.
6. Click "Install" to initiate the installation process.
By following these steps, users can successfully install the InRelate package and begin utilizing its functionalities within RStudio.

## 1.2 Running the Package in Rstudio
There are four main functions that users need to execute for pairwise relatedness calculation using the InRelate package. The initial three functions are designed to read in the three datasets required for genetic estimation.
1. readIndivq - Reads in admixture proportions obtained from ADMIXTURE/STRUCTURE

```{r}
#readIndivq("your indivq file")
#in my case, the example file from GitHub repository is called "example.indivq"

readIndivq("example.indivq")

#you can also store this into a variable
x = readIndivq("example.indivq")
```

2. readPkla - Reads in allele frequency data obtained from ADMIXTURE/STRUCTURE
```{r}
#readPkla("your pkla file")
#in my case, the example file from GitHub repository is called "examplepkla.txt"

readPkla("examplepkla.txt")

#you can also store this into a variable
y = readPkla("examplepkla.txt")
```

3. readStr - Reads in multi-locus genomic data in STRUCTURE format
```{r}
#readStr("your str file")
#in my case, the example file from GitHub repository is called "example.str"

readStr("example.str")

#you can also store this into a variable
z = readStr("example.str")
```

The last function is the most important function as it caluclates pairwise estimation.
4. computeRelatedness - Implements the InRelate estimator

There are two ways Rstudio can read this function. he first approach involves utilizing the variables where the data has been stored. Alternatively, the second method entails using the three read functions directly. The computeRelatedness function takes in three variables as parameters: the indivq file, the pkla file, and the str file.

```{r}
#this is utilizing the stored variables from previous example
computeRelatedness(x, y, z)

#this is utilizing the read functions directly
computeRelatedness(readIndivq("example.indivq"), readPkla("examplepkla.txt"), readStr("example.str"))

```

## 1.3 List of Functions in InRelate

Below is a list of functions used in InRelate, although users typically focus on the four main functions mentioned above:

getSubpop - Parses number of subpopulations from the INDIV file
getAlleles - Parses all unique alleles from PKLA file
getLoci - Parses genomic data from STR file
getIndiv - Parses admixture proportions from the INDIV file
computenumpairs - Computes the number of pairs using INDIV file
getIBS - Computes the IBS mode between pairs of alleles to compute pairwise relatedness
calcCoeff - Calculates IBD coefficients under all 9 IBS states
doOptimization - Implements optimization of the likelihood equation using Rsolnp package
calcRelatedness1 - Calculates relatedness with inbreeding
calcRelatedness2 - Calculates relatedness without inbreeding
Fst - Computes Fst from the Pegas package

# 2. Necessary Data (Input files)

To utilize InRelate, users must obtain ancestry information from genotype data using other ancestry inference tools such as ADMIXTURE, STRUCTURE, or MULTICLUST. For detailed instructions on running these packages, please refer to the original GitHub repository or respective package documentation for comprehensive user manuals.

These packages are designed to take commonly used genetic markers, including microsatellites, SNPs, and RFLPs. It also accept input in the form of .bed files, .ped files, or .geno files. If using .bed files, it is important to include the associated .bim and .fam files in the same directory. For .ped or .geno files, a corresponding PLINK .map file should also be in the same directory.

The output data from these tools may include various results, but the two key outputs of interest are the ancestry proportion data (INDIVQ files) and subpopulation allele frequency data (PKLA files).

## 2.1 Example Datasets

The example dataset and format for these datasets can be found on the Github Repository (https://github.com/itskhuyen/InRelate)

### 2.1.1 INDIVQ File

The INDIV file is an output file from ancestry inference tools like ADMIXTURE, STRUCTURE, or MULTICLUST. Each column in this dataset represents the membership coefficients for one individual. The second column provides a code number for the individual, while the fourth column provides the code number for the population to which the individual belongs. Columns 1, 3, and 5 are usually ignored.

Columns 6 onwards display the membership coefficients for clusters 1, 2, ..., K, respectively. Ideally, the numbers in these K columns should add up to 1.

### 2.1.2 PKLA File (need to fill out the headers for the data)
INDIV file is an outputs from ADMIXTURE, STRUCTURE, or MULTICLUST. 

### 2.1.3 STR File ((need to fill out the headers for the data))
STR file that contains the multilocus genomic data in STRUCTURE format.

## 2.3 Inrelate Outputs

InRelate generates two output files, namely the .relat file and the .deltas file. The .relat file is crucial for analyzing the relatedness dataset. It comprises three columns, with the first column indicating the relationship between two individuals. For instance, if it is "1_2," the subsequent values in column 2 and 3 represent the relatedness estimation for individual 1 and individual 2.

On the other hand, the .deltas file provides the calculation for the probability of the 9 different Identity by Descent (IBD) states. This probability is then used to compute the overall relatedness between individuals. It is not crucial for users data analysis process.

#3. Conclusion

InRelate stands out as one of the more accurate methods available today for relatedness estimation. Our research findings indicate that certain commonly used estimators such as those from COANCESTRY, tend to significantly overestimate or underestimate relatedness between half-siblings (HS) and full-siblings (FS). In contrast, InRelate and InRelateIB (which considers the presence of inbreeding within K subpopulations), demonstrate higher accuracy and lower variance across the overall dataset.


